Extract used propTypes detection

alexzherdev opened the pull request:
~~Pushing for visibility.
This is failing one existing test, which will get fixed after #1939 is merged (also will need a rebase).~~

Resolves #963, resolves #1871.

ljharb commented on lib/util/usedPropTypes.js:
```diff
@@ -0,0 +1,520 @@
+    'Program:exit': function() {
+      const list = components.list();
+
+      for (const component in list) {
```
`for..in` should be avoided; can this use normal array iteration methods?

alexzherdev replied:
Sure. `Object.keys(list).forEach` perhaps?
There's about a dozen other files already doing the same, handle them all?

ljharb replied:
yes, that would be great!

alexzherdev replied:
already done in https://github.com/yannickcr/eslint-plugin-react/pull/1946/commits/b35ea0c72317d4123237af32759f60858343b682 😄

alexzherdev commented:
I went ahead and fixed other `for..in` occurrences; I can go back and undo those if necessary. Also rebased.
`prop-types` is 201 line now, `no-unused-prop-types` is 144 🎉

ljharb commented on lib/rules/boolean-prop-naming.js:
```diff
@@ -248,7 +247,7 @@ module.exports = {
           }
 
-          if (!has(list, component) || (list[component].invalidProps || []).length) {
+          if ((list[component].invalidProps || []).length) {
```
maybe `if (list[component].invalidProps && list[component].invalidProps.length > 0)`?

ljharb commented on lib/rules/display-name.js:
```diff
@@ -216,12 +215,12 @@ module.exports = {
-            continue;
+        Object.keys(list).forEach(component => {
+          if (list[component].hasDisplayName) {
+            return;
```
`Object.keys(list).filter(component => !list[component].hasDisplayName).forEach`?

ljharb commented on lib/rules/no-multi-comp.js:
```diff
@@ -61,15 +60,15 @@ module.exports = {
-        for (const component in list) {
-          if (!has(list, component) || isIgnored(list[component]) || ++i === 1) {
-            continue;
+        Object.keys(list).forEach(component => {
```
`(component, i)` and then we don't need to `++` it, nor have `let i = 0` above.

alexzherdev replied:
Can't do this because we only want to `++i` if `isIgnored` is false and doesn't short-circuit. I.e. we want to start reporting from the _second_ non-ignored component.

alexzherdev replied:
Actually found a way to achieve this, also using `filter`.

ljharb commented on lib/rules/no-set-state.js:
```diff
@@ -74,12 +73,12 @@ module.exports = {
-        for (const component in list) {
-          if (!has(list, component) || isValid(list[component])) {
-            continue;
+        Object.keys(list).forEach(component => {
```
`Object.keys(list).filter(component => !isValid(list[component])).forEach`?

ljharb commented on lib/rules/no-unused-prop-types.js:
```diff
@@ -490,104 +124,20 @@ module.exports = {
-        for (const component in list) {
-          if (!has(list, component) || !mustBeValidated(list[component])) {
-            continue;
+        Object.keys(list).forEach(component => {
```
`Object.keys(list).filter(component => mustBeValidated(list[component])).forEach`?

ljharb commented on lib/rules/prop-types.js:
```diff
@@ -478,76 +181,20 @@ module.exports = {
-        for (const component in list) {
-          if (!has(list, component) || !mustBeValidated(list[component])) {
-            continue;
+        Object.keys(list).forEach(component => {
```
`Object.keys(list).filter(component => mustBeValidated(list[component])).forEach`?

ljharb commented on lib/rules/require-default-props.js:
```diff
@@ -626,21 +625,17 @@ module.exports = {
-            continue;
-          }
-
+        Object.keys(list).forEach(component => {
```
`Object.keys(list).filter(component => list[component].propTypes).forEach`?

ljharb commented on lib/rules/require-optimization.js:
```diff
@@ -221,12 +220,12 @@ module.exports = {
-        for (const component in list) {
-          if (!has(list, component) || list[component].hasSCU) {
-            continue;
+        Object.keys(list).forEach(component => {
```
`Object.keys(list).filter(component => !list[component].hasSCU).forEach`

ljharb commented on lib/util/Components.js:
```diff
@@ -123,9 +124,9 @@ class Components {
-    for (const i in this._list) {
-      if (!has(this._list, i) || this._list[i].confidence >= 2) {
-        continue;
+    Object.keys(this._list).forEach(i => {
```
`Object.keys(this._list).filter(i => this._list[i].confidence < 2).forEach`?

ljharb commented on lib/util/Components.js:
```diff
@@ -142,21 +143,22 @@ class Components {
-    for (const j in this._list) {
-      if (!has(this._list, j) || this._list[j].confidence < 2) {
-        continue;
+    Object.keys(this._list).forEach(j => {
```
`Object.keys(this._list).filter(j => this._list[j].confidence >= 2).forEach`?

ljharb commented on lib/util/Components.js:
```diff
@@ -697,7 +692,9 @@ function componentRule(rule, context) {
-  const allKeys = new Set(Object.keys(detectionInstructions).concat(Object.keys(propTypesInstructions)));
+  const usedPropTypesInstructions = usedPropTypesUtil(context, components, utils);
+  const allKeys = new Set(Object.keys(detectionInstructions).concat(Object.keys(propTypesInstructions))
+    .concat(Object.keys(usedPropTypesInstructions)));
```
`new Set(Object.keys(detectionInstructions).concat(Object.keys(propTypesInstructions), Object.keys(usedPropTypesInstructions)))`?

ljharb commented on lib/util/usedPropTypes.js:
```diff
@@ -0,0 +1,526 @@
+ * @returns {Boolean} True if the component must be validated, false if not.
+ */
+function mustBeValidated(component) {
+  return Boolean(
```
let's use `!!` instead of `Boolean(`

ljharb commented on lib/util/usedPropTypes.js:
```diff
@@ -0,0 +1,526 @@
+    let scope = context.getScope();
+    while (scope) {
+      if (
+        scope.block && scope.block.parent &&
```
if we cant get the whole condition on one line, let's put each condition on its own line

ljharb commented on lib/util/usedPropTypes.js:
```diff
@@ -0,0 +1,526 @@
+
+        if (LIFE_CYCLE_METHODS.indexOf(name) >= 0) {
+          return true;
+        } else if (checkAsyncSafeLifeCycles && ASYNC_SAFE_LIFE_CYCLE_METHODS.indexOf(name) >= 0) {
```
no need for the `else`

ljharb commented on lib/util/usedPropTypes.js:
```diff
@@ -0,0 +1,526 @@
+
+    if (node.kind === 'constructor') {
+      return true;
+    } else if (LIFE_CYCLE_METHODS.indexOf(nodeKeyName) >= 0) {
```
also here (ie, the `no-else-return` eslint rule)

alexzherdev replied:
yeah, I wonder why this isn't warning

ljharb commented on lib/util/usedPropTypes.js:
```diff
@@ -0,0 +1,526 @@
+    switch (type) {
+      case 'direct':
+        // Ignore Object methods
+        if (Object.prototype[name]) {
```
`name in Object.prototype` would be better, so as not to trigger getters

ljharb commented on lib/util/usedPropTypes.js:
```diff
@@ -0,0 +1,526 @@
+      // method -- mark it for used props.
+      if (isNodeALifeCycleMethod(node.parent.parent)) {
+        node.properties.forEach((property, i) => {
+          if (i === 0) {
```
this is just:
```js
if (node.properties.length > 0) {
  markPropTypesAsUsed(node.parent);
} 
```
the forEach seems useless?

ljharb commented on lib/util/usedPropTypes.js:
```diff
@@ -0,0 +1,526 @@
+    'Program:exit': function() {
+      const list = components.list();
+
+      Object.keys(list).forEach(component => {
```
`Object.keys(list).filter(component => mustBeValidated(list[component])).forEach`?

ThiefMaster commented:
I believe this change caused #2017.

The old logic was:

```js
    function isPropTypesUsage(node) {
      const isClassUsage = (
        (utils.getParentES6Component() || utils.getParentES5Component()) &&
        node.object.type === 'ThisExpression' && node.property.name === 'props'
      );
      const isStatelessFunctionUsage = node.object.name === 'props' && !isAssignmentToProp(node);
      const isNextPropsUsage = node.object.name === 'nextProps' && (inComponentWillReceiveProps() || inShouldComponentUpdate());
      return isClassUsage || isStatelessFunctionUsage || isNextPropsUsage;
}
```

and it got replaced with:

```js
  function isPropTypesUsage(node) {
    const isClassUsage = (
      (utils.getParentES6Component() || utils.getParentES5Component()) &&
      ((node.object.type === 'ThisExpression' && node.property.name === 'props')
      || isPropArgumentInSetStateUpdater(node))
    );
    const isStatelessFunctionUsage = node.object.name === 'props' && !isAssignmentToProp(node);
    return isClassUsage || isStatelessFunctionUsage || inLifeCycleMethod();
  }
```
